src/runtime/EncodedValue.h

```diff
@@ -218,17 +223,31 @@ class EncodedValue {
             return Value(value);
         }
 
-        PointerValue* v = (PointerValue*)m_data.payload;
-        if (((size_t)v) <= ValueLast) {
+        void* ptr = reinterpret_cast<void*>(m_data.payload);
+        if (((size_t)ptr) <= ValueLast) {
 #ifdef ESCARGOT_32
-            return Value(Value::FromTag, ~((uint32_t)v));
+            return Value(Value::FromTag, ~((uint32_t)ptr));
 #else
-            return Value(v);
+            return Value(reinterpret_cast<PointerValue*>(ptr));
 #endif
-        } else if (v->isDoubleInEncodedValue()) {
-            return Value(v->asDoubleInEncodedValue()->value());
         }
-        return Value(v);
+
+#ifdef ESCARGOT_32
+        size_t tag = readTagFromPointer(ptr);
+        if (tag == PointerValue::g_objectTag) {
+            return Value(reinterpret_cast<Object*>(ptr));
+        } else if (static_cast<unsigned char>(tag) == ENCODED_VALUE_DOUBLE_TAG_IN_DATA) {
+            return Value(Value::EncodeAsDouble, reinterpret_cast<DoubleInEncodedValue*>(ptr)->value());
```

Comment:
One more question.
Is there any specific reason that object tag check is done only for 32bit case?
It seems that `PointerValue::g_objectTag` is used only for this case.

Reply:
Only 32-bit tests pointer is object pointer.
We save tag as ObjectPointer or OtherPointer
It makes Value::isObject() faster

Reply:
(AFAIK) `PointerValue::g_objectTag` is the vtable address of `Object`, and `ObjectPointerTag` tag stored in Value (32bit) is different.
IMHO we don't need to check `g_objectTag` here because `Value(reinterpret_cast<PointerValue*>(ptr))` would do the identical check insde. So, I think that its possible to merge each 32 bit and 64 bit operations here.

Reply:
the first word is "ENCODED_VALUE_DOUBLE_TAG_IN_DATA" or vtable address
and the second word is PointerValueTags.
If I remove g_objectTag(vptr), we should next word of pointer.
I am preparing next pr for moving ENCODED_VALUE_DOUBLE_TAG_IN_DATA to next word of the pointer like PointerValues

Reply:
oh, I see